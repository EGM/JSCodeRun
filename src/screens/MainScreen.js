import React, { Component } from 'react';
import {
  StyleSheet,
  View,
  Keyboard,
  TouchableWithoutFeedback,
  KeyboardAvoidingView,
  StatusBar
} from 'react-native';
import CodeEditor from '../components/CodeEditor';
import Output from '../components/Output';
import HeaderBar from '../components/HeaderBar';
import defaultCode from '../constants/code/hello-world';

const styles = StyleSheet.create({
  container: {
    flex:1,
    backgroundColor: '#777'
  },
  screenContainer: {
    flex: 1
  },
  editorWrapper: {
    flex: 2,
    alignSelf: 'stretch'
  },
  outputWrapper: {
    flex: 1,
    alignSelf: 'stretch',
    backgroundColor: '#fff'
  }
});
  

export default class MainScreen extends Component {
  constructor(props) {
    super(props);
    this.state = {
      code: defaultCode,
      output: []
    };
  }

  handleCodeChange = code => {
    this.setState({
      code
    });
  };

  onClearOutput = () => {
    this.setState({
      output: []
    });
  }

  runCode = () => {
    Keyboard.dismiss();
    const output = [];
    const consoleTemp = Object.assign({}, console);

    console.log = message => {
      output.push({
        message,
        status: 'OK'
      });
      //log(arguments);
    };
    console.error = message => {
      output.push({
        message,
        status: 'ERROR'
      });
    };
    try {
      new Function(this.state.code)();
    } catch (err) {
      output.push({
        message: err.toString(),
        status: 'ERROR'
      });
    }

    console = consoleTemp;
    this.setState({
      output
    });
  };

  render() {
    return (
      <View style={styles.container}>
        <StatusBar
            barStyle="light-content"
            animated
            translucent
        />
        <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
          <View style={styles.screenContainer}>
            <HeaderBar runCode={this.runCode} generateCode={(code) => this.setState({code})}/>
            <KeyboardAvoidingView behavior="height" style={styles.editorWrapper}>
              <CodeEditor
                code={this.state.code}
                handleCodeChange={this.handleCodeChange}
              />
            </KeyboardAvoidingView>
            <View style={styles.outputWrapper}>
              <Output
                output={this.state.output}
                onClearOutput={this.onClearOutput} />
            </View>
          </View>
        </TouchableWithoutFeedback>
      </View>
    );
  }
}