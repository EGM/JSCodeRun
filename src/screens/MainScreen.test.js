import React, { Component } from 'react';
import renderer from 'react-test-renderer';
import MainScreen from './MainScreen';

jest.mock('../components/CodeEditor', () => 'CodeEditor');
jest.mock('../components/Output', () => 'Output');
jest.mock('../components/HeaderBar', () => 'HeaderBar');

describe('MainScreen', () => {
  it('should render correctly', () => {
    const rendered = renderer.create(
      <MainScreen />
    ).toJSON();
    expect(rendered).toMatchSnapshot();
  });
})