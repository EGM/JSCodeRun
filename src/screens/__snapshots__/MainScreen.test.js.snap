// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScreen should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#777",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 2,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={undefined}
        testID={undefined}
      >
        <HeaderBar
          generateCode={[Function]}
          runCode={[Function]}
        />
        <View
          keyboardVerticalOffset={0}
          onLayout={[Function]}
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        >
          <CodeEditor
            code="// Delete me and write some code!
          function sayHello() {
            console.log(\\"Hello World!\\");
          }
          
          sayHello();"
            handleCodeChange={[Function]}
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "backgroundColor": "#fff",
          "flex": 1,
        }
      }
    >
      <Output
        onClearOutput={[Function]}
        output={Array []}
      />
    </View>
  </View>
</View>
`;
